### PR TITLE
feat(framework) Add selectors to stop command

### LINKED ISSUE #7945: Allow "semantic parameters" for `flwr stop` (such as `flwr stop latest`)
### Type

Feature

### Description

As I'm getting started with Flower and tinkering with my experiment setup, I find myself needing to stop and restart runs a lot. As-is, it seems the cycle goes like so:
1. Run the command `flwr run` to start the run.
2. Once I want to stop the run, run the command `flwr list` to get the list of current runs.
3. In the list, copy the top Run ID.
4. Run the command `flwr stop [RUN_ID]` to actually stop the run.

While I recognize it certainly won't be the case for everyone, I pretty much always only intend to have one run going at a time on my local machine, and I imagine there must be a fair number of other users who would be doing the same thing.

As such, I think it could be useful to add semantic parameters (I think that'd be an accurate term!) for the `stop` command. Specifically:
- `flwr stop latest` could fetch the Run ID of the most recent run and stop it (equivalent to the process I described above).
- `flwr stop all` could fetch Run IDs of _all currently-running runs_ and stop them.

### Planned Implementation

I am new to Flower's codebase, so I'd have to do a bit more studying and experimenting to make sure. But from looking at the code right now, I think this would just involve modifying `flwr/cli/stop.py`.
- The `run_id` argument of the `stop()` function could take _either_ an `int` (for a run ID) or a `str` (for a semantic parameter like `latest` or `all`).
- If one of the semantic parameters is detected, before calling `_stop_run()` either the most recent run's ID could be found, or all active runs' IDs could be found and `_stop_run()` could be called on all of them in sequence.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by danielpolimac**
### Description

Stopping a recent run currently requires listing runs, copying an ID, and passing that numeric ID to `flwr stop`. Stopping every active run requires repeating the process.

### Related issues/PRs

Fixes #7945.

### Explanation

Allow the `flwr stop` argument to be either a numeric run ID or one of two selectors:

- `latest` stops the newest unfinished run.
- `all` stops all unfinished runs, newest first.

Unfinished includes pending, starting, and running runs, which ensures a pending run can also be stopped. Existing numeric-ID behavior remains unchanged. The JSON response for `all` contains a single `run-ids` list instead of emitting multiple JSON documents.

Arguments are validated before configuration migration or connection setup. Batch execution attempts every selected run before reporting partial failures. A run selected by either `latest` or `all` that finishes concurrently is treated as already satisfying the stop request, including when the stop RPC returns an unsuccessful response instead of an API error. Unrelated failures remain visible after the rest of an `all` batch is attempted.

Add focused tests for numeric IDs, active-run filtering and ordering, real CLI validation before setup, negative IDs, the no-active-runs case, both concurrent-completion response forms, batch failure continuation, RPC order, and JSON output.

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [ ] Update documentation (the CLI help text documents the new argument values)
- [x] Address LLM-reviewer comments, if applicable
- [x] Make local checks pass
- [ ] Make CI checks pass
- [ ] Ping maintainers on Slack (channel `#contributions`)

### Any other comments?

Local verification:

- `python -m pytest py/flwr/cli/stop_test.py py/flwr/cli/cli_test.py -k stop`
- `python -m ruff check py/flwr/cli/stop.py py/flwr/cli/stop_test.py --no-respect-gitignore`
- `python -m mypy py/flwr/cli/stop.py py/flwr/cli/stop_test.py`
- `python -m pylint py/flwr/cli/stop.py py/flwr/cli/stop_test.py`
- `uv run --no-sync --python=3.11.14 ./dev/test.sh false`

### CODE REVIEW

File: framework/py/flwr/cli/stop.py [hunk 1]
```diff
@@ -84,14 +86,66 @@ def stop(  # pylint: disable=R0914
...
-            typer.secho(f"✋ Stopping run ID {run_id}...", fg=typer.colors.GREEN)
-            _stop_run(stub=stub, run_id=run_id, is_json=is_json)
+            run_ids = _resolve_run_ids(stub, run_id)
+            stop_all = run_id.lower() == "all"
+            for resolved_run_id in run_ids:
+                typer.secho(
+                    f"✋ Stopping run ID {resolved_run_id}...",
+                    fg=typer.colors.GREEN,
+                )
+                _stop_run(
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py]: Addressed in 8269b36e1: batch stops now continue after per-run failures and report all partial failures after attempting every selected run. The new regression test verifies the later run is still attempted.

File: framework/py/flwr/cli/stop.py [hunk 2]
```diff
@@ -84,14 +86,66 @@ def stop(  # pylint: disable=R0914
...
+                    stub=stub,
+                    run_id=resolved_run_id,
+                    is_json=is_json and not stop_all,
+                )
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py]: Addressed in 8269b36e1: `all` catches each per-run CLI failure, continues through the batch, and reports collected failures afterward.

File: framework/py/flwr/cli/stop.py [hunk 3]
```diff
@@ -84,14 +86,66 @@ def stop(  # pylint: disable=R0914
...
+            if is_json and stop_all:
+                print_json_to_stdout(
+                    {
+                        "success": True,
+                        "run-ids": [
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py]: Addressed in 8269b36e1: command-level coverage now invokes `stop(..., "all", --format json)`, verifies newest-first `StopRun` calls, asserts a single JSON document, and separately covers continuation after a failed stop.

File: framework/py/flwr/cli/stop.py [hunk 4]
```diff
@@ -84,14 +86,74 @@ def stop(  # pylint: disable=R0914
...
+    """Stop resolved run IDs and display the result."""
+    failures = []
+    for run_id in run_ids:
+        typer.secho(f"✋ Stopping run ID {run_id}...", fg=typer.colors.GREEN)
+        try:
+            _stop_run(stub=stub, run_id=run_id, is_json=is_json and not stop_all)
+        except click.ClickException as err:
+            if not stop_all:
+                raise
+            failures.append(f"Run {run_id}: {err.format_message()}")
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py:167]: Addressed in 8ed71b558: `RUN_ALREADY_FINISHED` is now recognized by its typed Flower API error code and treated as a successful/skippable outcome only for `all`. Other errors still remain visible after the batch. A serialized gRPC-error regression test covers the race.

File: framework/py/flwr/cli/stop.py [hunk 5]
```diff
@@ -84,15 +88,85 @@ def stop(  # pylint: disable=R0914
             channel = init_channel_from_connection(superlink_connection)
             stub = ControlStub(channel)  # pylint: disable=unused-variable # noqa: F841
 
-            typer.secho(f"✋ Stopping run ID {run_id}...", fg=typer.colors.GREEN)
-            _stop_run(stub=stub, run_id=run_id, is_json=is_json)
+            run_ids = _resolve_run_ids(stub, run_id)
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py]: Addressed in 682dedbb5: `_parse_run_id` now runs as the first operation inside the output handler, before warnings, migration, configuration reads, or channel setup. A real Typer invocation with `flwr stop lates` asserts that migration and channel initialization are never called.

File: framework/py/flwr/cli/stop.py [hunk 6]
```diff
@@ -84,15 +88,85 @@ def stop(  # pylint: disable=R0914
...
+    """Stop resolved run IDs and display the result."""
+    failures = []
+    for run_id in run_ids:
+        typer.secho(f"✋ Stopping run ID {run_id}...", fg=typer.colors.GREEN)
+        try:
+            _stop_run(
+                stub=stub,
+                run_id=run_id,
+                is_json=is_json and not stop_all,
+                ignore_finished=stop_all,
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py]: Addressed in 682dedbb5: already-finished handling is now enabled for both selectors (`latest` and `all`) while numeric-ID behavior remains unchanged. A dedicated `latest` race regression test covers the API-error path.

File: framework/py/flwr/cli/stop.py [hunk 7]
```diff
@@ -103,9 +177,28 @@ def _stop_run(stub: ControlStub, run_id: int, is_json: bool) -> None:
...
+        flower_error = FlowerError.from_json(details)
+        if (
+            ignore_finished
+            and flower_error is not None
+            and flower_error.code == ApiErrorCode.RUN_ALREADY_FINISHED
+        ):
+            raise _RunAlreadyFinishedError
+
+    try:
+        with flwr_cli_grpc_exc_handler(custom_handler=raise_if_already_finished):
```
[the remark this answers is not in the document]
Comment (danielpolimac) [reply framework/py/flwr/cli/stop.py:212]: Addressed in 682dedbb5: an unsuccessful stop response for a selector now triggers `ListRuns(run_id=...)`; it is accepted only when the run is confirmed finished, otherwise the original failure remains. The regression test covers the `all` race path.
